### DISCUSSION TITLE
Prisma do not create new connections in the pool

### DISCUSSION THREAD

**Discussion by HalexV** [prisma/prisma#21208] [Q&A]
### Bug description

During a gatling benchmark test when the quantity of requests is very high, the amount of opened connections in the pool don't increase to support the load test.

The scenario:
- A Nestjs app;
- Using PostgreSQL;
- Cluster mode true with 2 cluster workers;
- And 2 app instances;
- Total of 4 apps with connection limit set to 60;
- A Postgres container;
- A NGINX container redirecting requests to the apps;
- A pgHero container to get statistics about Postgres database;

The containers are restricted to use together 1.5 CPUs and 3.0GB ram.

Below is the error of can't fetch connection, even though the total of expected opened connections to be 240.

![image](https://github.com/prisma/prisma/assets/14897195/5e9b333e-5e0c-407f-8909-5b7a1ccb6a7b)

### How to reproduce

Steps to up the containers:

1. Clone https://github.com/HalexV/rinha-de-backend-2023-q3
2. cd rinha-de-backend-2023-q3
3. Run npm ci
4. Run npm run build
5. Run docker compose up

With containers up, steps to check opened database connections (pgHero):
1. On browser access localhost:80
2. Click on connections
3. You need to refresh this page to see the quantity of connections changing

With containers up, steps to run the gatling test:
1. cd gatling
2. run ./install-gatling
3. run mkdir user-files/resources
4. run ./geradores/faker/gerar-pessoas > ./user-files/resources/pessoas-payloads.tsv
5. run ./geradores/faker/gerar-termos-busca > ./user-files/resources/termos-busca.tsv
6. The steps above you do only once
7. run ./run-test to run the gatling test
8. While the test runs you need to reload connections page on pgHero

### Expected behavior

On pgHero connections, the connections amount should be 4 * 60 or 240 connections at high load test.

### Prisma information

```prisma
// schema.prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["postgresqlExtensions"]
}

datasource db {
  provider   = "postgresql"
  url        = env("DATABASE_URL")
  extensions = [pg_trgm]
}

model Person {
  id         String   @id @default(uuid())
  apelido    String   @unique @db.VarChar(32)
  nome       String   @db.VarChar(100)
  nascimento DateTime
  stack      String?
  searchable String?

  @@index([searchable(ops: raw("gist_trgm_ops (siglen='64')"))], type: Gist)
  @@map("people")
}

```

```ts
// prisma.service
import { Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common'
import { PrismaClient } from '@prisma/client'

@Injectable()
export class PrismaService
  extends PrismaClient
  implements OnModuleInit, OnModuleDestroy
{
  constructor() {
    super({
      log: ['info', 'error', 'warn'],
    })
  }

  async onModuleInit() {
    return await this.$connect()
  }

  async onModuleDestroy() {
    return await this.$disconnect()
  }
}

// prisma-people-repository.ts
import { PeopleRepository } from '@/domain/my-stacks/application/repositories/people-repository'
import { Person } from '@/domain/my-stacks/enterprise/entities/person'
import { PrismaService } from '../prisma.service'
import { PrismaPersonMapper } from '../mappers/prisma-person-mapper'
import { Injectable } from '@nestjs/common'

@Injectable()
export class PrismaPersonRepository implements PeopleRepository {
  constructor(private prisma: PrismaService) {}

  async create(person: Person): Promise<void> {
    const data = PrismaPersonMapper.toPrisma(person)

    await this.prisma.person.create({
      data,
    })
  }

  async findById(id: string): Promise<Person | null> {
    const person = await this.prisma.person.findUnique({
      where: {
        id,
      },
    })

    if (!person) {
      return null
    }

    return PrismaPersonMapper.toDomain(person)
  }

  async findByApelido(apelido: string): Promise<Person | null> {
    const person = await this.prisma.person.findUnique({
      where: {
        apelido,
      },
    })

    if (!person) {
      return null
    }

    return PrismaPersonMapper.toDomain(person)
  }

  async findManyByQuery(query: string): Promise<Person[]> {
    const people = await this.prisma.person.findMany({
      where: {
        searchable: {
          contains: query,
          mode: 'insensitive',
        },
      },
      take: 50,
    })

    return people.map(PrismaPersonMapper.toDomain)
  }

  async count(): Promise<number> {
    return await this.prisma.person.count()
  }
}

```

### Environment & setup

- OS: 
  - Xubuntu
  - Distributor ID: Ubuntu
  - Description:    Ubuntu 22.04.3 LTS
  - Release:        22.04
  - Codename:       jammy
- Database: PostgreSQL
- Node.js version: 18;

### Prisma Version

```
prisma                  : 5.2.0
@prisma/client          : 5.2.0
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine 2804dc98259d2ea960602aca6b8e7fdc03c1758f (at node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli 2804dc98259d2ea960602aca6b8e7fdc03c1758f (at node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.2.0-25.2804dc98259d2ea960602aca6b8e7fdc03c1758f
Default Engines Hash    : 2804dc98259d2ea960602aca6b8e7fdc03c1758f
Studio                  : 0.494.0
Preview Features        : postgresqlExtensions
```

**Comment by janpio** [comment]
> On pgHero connections, the connections amount should be 4 * 60 or 240 connections at high load test.

And it currently is not? I could not extract that information from your post @HalexV.

**Reply by HalexV** [reply]
@janpio Yes, it currently is not.

Well, I started to use pgHero to check queries delays of this application. The queries on db are fast.

Even setting a connection limit higher, I keep getting the fetch connection timeout.

So, I decided to check the amount of opened connections to the database through pgHero connection page.

With cluster mode off and 2 app instances with connection limit set to 50, on high load test the quantity of connections goes to 100, but if I set connection limit to 140 and run the same test, the apps cannot increase the amount of connections.

When the apps can increase your connections to db, the usage amount of CPU and RAM don't increase much.

So, I don't know why this is happening.

**Reply by janpio** [reply]
Can you be more precise, what you refer to with "connection limit" above? That could either be the connection limit on your database, or the connection pool size, `connection_limit`, setting of Prisma Client via the connection string.

I understand you are seeing connection pool timeouts from Prisma, which you do not seem to expect.

**Reply by janpio** [reply]
Ok, I followed your instructions from your initial post and I can see that while running the Gatling test, after some time there start to happen a lot of the timeouts that you describe. The pgHero connections overview only showed ~20 connections to be in use.

The setup is pretty complex though, so I'll have to ask some more questions:
- What exactly does the Gatling test setup do and run? (I know almost nothing about Gatling, so a quick description of the whole setup there would be nice)
- How can I experience the same that Gatling is doing manually?
- Where is the actual code that is being executed?

One thing that will definitely be interesting and helpful is to have an endpoint in the app that you can manually call that returns the Json output of the metrics preview feature: https://www.prisma.io/docs/concepts/components/prisma-client/metrics That way, similar to the pgHero connection overview, you can get an idea what is happening inside Prisma (E.g. how many queries, how many connection and so on).

**Reply by HalexV** [reply]
> Can you be more precise, what you refer to with "connection limit" above? That could either be the connection limit on your database, or the connection pool size, connection_limit, setting of Prisma Client via the connection string.

`connection_limit` I'm referred to connection pool size setting of Prisma Client via the connection string.

> What exactly does the Gatling test setup do and run? (I know almost nothing about Gatling, so a quick description of the whole setup there would be nice)

1. You downloaded the gatling program with the install-gatling shell script. Gatling works similar to Autocannon. It uses a simulation file that contains settings to which http calls to do and how many simultaneous users It will generate to make this http calls. The simulation file used is in `gatling/user-files/simulations/rinhabackend/RinhaBackendSimulation.scala`. The simulation file was created for this test.
2. The payload for requests that this simulation use It was generated in `gatling/user-files/resources/*.tsv`.
3. The run-test script just run gatling with the simulation file. At this moment, Gatling starts to make http calls to the apps, similar to Autocannon.
4. Gatling while making requests, It starts to collect metrics about success requests, failed requests, delay of requests, etc.
5. When Gatling finishes, It generate a HTML report in `gatling/user-files/results`

> How can I experience the same that Gatling is doing manually?

Manually is not possible, just programmatically using something like Autocannon.

> Where is the actual code that is being executed?

At `gatling/user-files/simulations/rinhabackend/RinhaBackendSimulation.scala`.

> One thing that will definitely be interesting and helpful is to have an endpoint in the app that you can manually call that returns the Json output of the metrics preview feature.

I will do it and I will check it out.

**Reply by HalexV** [reply]
I added prisma metrics at /prisma/metrics route. Update your local branch.

I did a test and I collected this prisma metric during the higher load:

```json
{
	"counters": [
		{
			"key": "prisma_client_queries_total",
			"labels": {
				"appId": "101:1"
			},
			"value": 8713,
			"description": "Total number of Prisma Client queries executed"
		},
		{
			"key": "prisma_datasource_queries_total",
			"labels": {
				"appId": "101:1"
			},
			"value": 8726,
			"description": "Total number of Datasource Queries executed"
		},
		{
			"key": "prisma_pool_connections_open",
			"labels": {
				"appId": "101:1"
			},
			"value": 3,
			"description": "Number of currently open Pool Connections"
		}
	],
	"gauges": [
		{
			"key": "prisma_client_queries_active",
			"labels": {
				"appId": "101:1"
			},
			"value": 0,
			"description": "Number of currently active Prisma Client queries"
		},
		{
			"key": "prisma_client_queries_wait",
			"labels": {
				"appId": "101:1"
			},
			"value": 1898,
			"description": "Number of queries currently waiting for a connection"
		},
		{
			"key": "prisma_pool_connections_busy",
			"labels": {
				"appId": "101:1"
			},
			"value": 2,
			"description": "Number of currently busy Pool Connections (executing a database query)"
		},
		{
			"key": "prisma_pool_connections_idle",
			"labels": {
				"appId": "101:1"
			},
			"value": 58,
			"description": "Number of currently unused Pool Connections (waiting for the next pool query to run)"
		},
		{
			"key": "prisma_pool_connections_opened_total",
			"labels": {
				"appId": "101:1"
			},
			"value": 3,
			"description": "Total number of Pool Connections opened"
		}
	],
	"histograms": [
		{
			"key": "prisma_client_queries_duration_histogram_ms",
			"labels": {
				"appId": "101:1"
			},
			"value": {
				"buckets": [
					[
						0,
						0
					],
					[
						1,
						8381
					],
					[
						5,
						266
					],
					[
						10,
						8
					],
					[
						50,
						27
					],
					[
						100,
						0
					],
					[
						500,
						2
					],
					[
						1000,
						8
					],
					[
						5000,
						15
					],
					[
						50000,
						4
					]
				],
				"sum": 69437.87904600003,
				"count": 8711
			},
			"description": "Histogram of the duration of all executed Prisma Client queries in ms"
		},
		{
			"key": "prisma_client_queries_wait_histogram_ms",
			"labels": {
				"appId": "101:1"
			},
			"value": {
				"buckets": [
					[
						0,
						0
					],
					[
						1,
						8741
					],
					[
						5,
						1
					],
					[
						10,
						0
					],
					[
						50,
						0
					],
					[
						100,
						0
					],
					[
						500,
						3
					],
					[
						1000,
						7
					],
					[
						5000,
						31
					],
					[
						50000,
						172
					]
				],
				"sum": 2167680.5833410006,
				"count": 8955
			},
			"description": "Histogram of the wait time of all queries in ms"
		},
		{
			"key": "prisma_datasource_queries_duration_histogram_ms",
			"labels": {
				"appId": "101:1"
			},
			"value": {
				"buckets": [
					[
						0,
						0
					],
					[
						1,
						8542
					],
					[
						5,
						117
					],
					[
						10,
						17
					],
					[
						50,
						18
					],
					[
						100,
						0
					],
					[
						500,
						4
					],
					[
						1000,
						7
					],
					[
						5000,
						15
					],
					[
						50000,
						6
					]
				],
				"sum": 80020.79812600007,
				"count": 8726
			},
			"description": "Histogram of the duration of all executed Datasource Queries in ms"
		}
	]
}
```

**Reply by janpio** [reply]
> At `gatling/user-files/simulations/rinhabackend/RinhaBackendSimulation.scala`.

Can you explain in your words what this is supposed to be doing?
If I understand, roughly, it is supposed to send a lot of HTTP requests (3 different kind) to http://localhost:9999?
Where exactly was that server on `:9999` started? How can I restart it before running the Gatling tests?

**Reply by HalexV** [reply]
> Can you explain in your words what this is supposed to be doing?

This file contains the setup to Gatling make the http requests.

Into the file (`/gatling/user-files/simulations/rinhabackend/RinhaBackendSimulation.scala`):

Set the baseURL to use to make requests and the name of the agent making the requests.
```scala
val httpProtocol = http
    .baseUrl("http://localhost:9999")
    .userAgentHeader("Agente do Caos - 2023")
```

Setup to make requests of creating and getting person.
```scala
// scenario: label to identify this set of requests.
val criacaoEConsultaPessoas = scenario("Criação E Talvez Consulta de Pessoas")
     // feed: load people payload to use to make requests of creating and getting person.
    .feed(tsv("pessoas-payloads.tsv").circular())
    // exec: execute the http request
     .exec(
      // the next line defines a label that it will use on the reports result.
      http("criação")
      // next line: method POST at /pessoas (create a person)
      .post("/pessoas").body(StringBody("#{payload}"))
      .header("content-type", "application/json")
      // next line check the response status. It will be used to count success requests on reports.
      .check(status.in(201, 422, 400))
      // next line save the response status and use it on checkIf to make the get person if status is 201 (person created)
      .check(status.saveAs("httpStatus"))
      .checkIf(session => session("httpStatus").as[String] == "201") {
        header("Location").saveAs("location")
      }
    )
    .pause(1.milliseconds, 30.milliseconds)
    // next line make the get person request based on status 201 collected before. If status was different of 201, this request is not made.
    .doIf(session => session.contains("location")) {
      exec(
        http("consulta")
        .get("#{location}")
      )
    }
```

Next setups to make requests of fetch people based on a query string and an invalid request that try to fetch people without pass the query string variable 't' and its value.
```scala
// Fetch people by the term passed on query string field 't'.
val buscaPessoas = scenario("Busca Válida de Pessoas")

// Fetch people without query string 't'. It will ever return 400.
val buscaInvalidaPessoas = scenario("Busca Inválida de Pessoas")
```

The setUp function set how many users will be simulated making requests using the scenarios defined above on the code and for how long.
```scala
setUp(
    // Here inject users that will use the scenario criacaoEConsultaPessoas (create and get person)
    criacaoEConsultaPessoas.inject(
      // First it will start with just 2 users making requests during 10 seconds.
      constantUsersPerSec(2).during(10.seconds), // warm up
     // Next, 5 users making requests during 15 seconds.
      constantUsersPerSec(5).during(15.seconds).randomized, // are you ready?
      // Finally, the hard part, start with 6 users and increase up to 600 making requests during 3 minutes.
      rampUsersPerSec(6).to(600).during(3.minutes) // lezzz go!!!
    ),
    // Setup to inject users using the buscaPessoas scenario (fetch people)
    buscaPessoas.inject(
      constantUsersPerSec(2).during(25.seconds), // warm up
      
      rampUsersPerSec(6).to(100).during(3.minutes) // lezzz go!!!
    ),
    // Setup to inject users using the buscaInvalidaPessoas scenario (fetch people, but without using the query string 't').
    buscaInvalidaPessoas.inject(
      constantUsersPerSec(2).during(25.seconds), // warm up
      
      rampUsersPerSec(6).to(40).during(3.minutes) // lezzz go!!!
    )
  ).protocols(httpProtocol)
```

> If I understand, roughly, it is supposed to send a lot of HTTP requests (3 different kind) to http://localhost:9999/? Where exactly was that server on :9999 started? How can I restart it before running the Gatling tests?

Yes and the server on :9999 is the nginx doind proxy reverse to the two apps. Nginx, two app instances, Postgre db and pgHero were up with docker compose command. You can restart the containers with `docker compose restart`.

**Reply by janpio** [reply]
Ok, I think roughly now what the Gatling setup is doing.

Hm, I can not get `/prisma/metrics` to return anything else than 404 on `:9999`. 
(Even if, it would of course only show a random of the two apps that are actually running Prisma)

Is the load balancer relevant for the current experiment? Would removing it not simplify the situation a lot?

**Reply by janpio** [reply]
With my increased understanding of the setup, took another look:

Is it correct that you run 2 application servers with 60 database connections in the connection pool each, so 120 queries can run in parallel, and then the Gatling load test runs 600+ constant users that make requests all the time, one after another over multiple minutes? 
If so, it is not really surprising that 120 connections can not handle 600 constant users, correct? Or am I missing something?

**Reply by HalexV** [reply]
> Hm, I can not get /prisma/metrics to return anything else than 404 on :9999.

I'm sorry, It's my fault. You need to do:
1. Update your branch with git pull
2. run `rm -rf dist/` to remove the builded app
3. run `docker image rm rinha-de-backend-2023-q3-app1:latest rinha-de-backend-2023-q3-app2:latest` to remove the images.
4. run `npm run build`
5. run `docker compose up`

 > Is the load balancer relevant for the current experiment? Would removing it not simplify the situation a lot?

It's not. I removed Nginx, the second app instance and set cluster false and prisma connection_limit to 240.

> Is it correct that you run 2 application servers with 60 database connections in the connection pool each, so 120 queries can run in parallel, and then the Gatling load test runs 600+ constant users that make requests all the time, one after another over multiple minutes?

Actually was 2 apps in cluster mode with 2 workers or simplifying 4 apps with 60 database connections each and which should be resulting in 240 opened connections to database.

Yes, it is correct. I'm doing a benchmark to see how much a nestjs app with prisma can process requests.

> If so, it is not really surprising that 120 connections can not handle 600 constant users, correct? Or am I missing something?

Actually should be 240 connections opened, but somehow these additional connections are not being opened. The question is that I can't check if 240 connections can handle 600 users if the 240 connections are not opened.

I updated the project and up one app with prisma connection_limit equals to 240. I tested again and I did not see the 240 connections opened.

How many computer resources prisma needs to open a new connection in the pool?

**Reply by janpio** [reply]
I think your Gatling test code is not doing what you think it should be doing. With `constantConcurrentUsers(10).during(1.minutes)` I get 10 or 11 connections opened in the connection pool.

With other setups I get a lot higher numbers:
<img width="618" alt="image" src="https://github.com/prisma/prisma/assets/183673/589af6a7-945b-4af4-9f55-bda95066744a">

```
{
	"counters": [{
		"key": "prisma_client_queries_total",
		"labels": {
			"appId": "187"
		},
		"value": 11380,
		"description": "Total number of Prisma Client queries executed"
	}, {
		"key": "prisma_datasource_queries_total",
		"labels": {
			"appId": "187"
		},
		"value": 11466,
		"description": "Total number of Datasource Queries executed"
	}, {
		"key": "prisma_pool_connections_open",
		"labels": {
			"appId": "187"
		},
		"value": 113,
		"description": "Number of currently open Pool Connections"
	}],
	"gauges": [{
		"key": "prisma_client_queries_active",
		"labels": {
			"appId": "187"
		},
		"value": 0,
		"description": "Number of currently active Prisma Client queries"
	}, {
		"key": "prisma_client_queries_wait",
		"labels": {
			"appId": "187"
		},
		"value": 46,
		"description": "Number of queries currently waiting for a connection"
	}, {
		"key": "prisma_pool_connections_busy",
		"labels": {
			"appId": "187"
		},
		"value": 71,
		"description": "Number of currently busy Pool Connections (executing a database query)"
	}, {
		"key": "prisma_pool_connections_idle",
		"labels": {
			"appId": "187"
		},
		"value": 169,
		"description": "Number of currently unused Pool Connections (waiting for the next pool query to run)"
	}, {
		"key": "prisma_pool_connections_opened_total",
		"labels": {
			"appId": "187"
		},
		"value": 113,
		"description": "Total number of Pool Connections opened"
	}],
	"histograms": [{
		"key": "prisma_client_queries_duration_histogram_ms",
		"labels": {
			"appId": "187"
		},
		"value": {
			"buckets": [
				[0, 0],
				[1, 751],
				[5, 998],
				[10, 898],
				[50, 1837],
				[100, 3675],
				[500, 2732],
				[1000, 271],
				[5000, 147],
				[50000, 0]
			],
			"sum": 1309987.0413359997,
			"count": 11309
		},
		"description": "Histogram of the duration of all executed Prisma Client queries in ms"
	}, {
		"key": "prisma_client_queries_wait_histogram_ms",
		"labels": {
			"appId": "187"
		},
		"value": {
			"buckets": [
				[0, 0],
				[1, 11426],
				[5, 1],
				[10, 0],
				[50, 0],
				[100, 0],
				[500, 0],
				[1000, 0],
				[5000, 0],
				[50000, 0]
			],
			"sum": 11.556984999999996,
			"count": 11427
		},
		"description": "Histogram of the wait time of all queries in ms"
	}, {
		"key": "prisma_datasource_queries_duration_histogram_ms",
		"labels": {
			"appId": "187"
		},
		"value": {
			"buckets": [
				[0, 0],
				[1, 978],
				[5, 848],
				[10, 915],
				[50, 1874],
				[100, 3712],
				[500, 2696],
				[1000, 278],
				[5000, 165],
				[50000, 0]
			],
			"sum": 1346264.3140390008,
			"count": 11466
		},
		"description": "Histogram of the duration of all executed Datasource Queries in ms"
	}]
}
```

> Yes, it is correct. I'm doing a benchmark to see how much a nestjs app with prisma can process requests.

That depends on the hardware it runs on. With a beefy machine, the database and Prisma will be able to handle a _lot_ of concurrent queries.

> How many computer resources prisma needs to open a new connection in the pool?

Almost none. Running the queries is what uses resources.

**Reply by HalexV** [reply]
I will replace prisma with pg driver on a new branch and check if something changes.

Thank you very much for your attention.
